Refactor code cache

qwe661234 opened the pull request:
According to the MIR author statement, MIR does not generate position-independent code. This means that the code can only be executed in the location where it was originally generated. Since it is not possible to move the generated code to a specific memory location, we need to refactor a code cache that solely stores the memory address of the generated code.

The following statistics demonstrate the performance improvement observed when importing JIT (Just-in-Time) compilation on various host machines.

* Core i7-8700

|Metric| rv32emu (interpreter) |rv32emu (JIT) |Speedup|
|--------| -------- | -------- | -------- | 
| CoreMark | 1155.174 (iter/s)| 1932.496 (iter/s) |+67.2％|
| dhrystone  | 1282 DMIPS |2495.11 DMIPS |+94.6%|

* eMAG 8180

|Metric| rv32emu (interpreter) |rv32emu (baseline JIT) |Speedup|
|--------| -------- | -------- | -------- | 
| CoreMark | 438.073 (iter/s)| 644.426 (iter/s) |+47.1％|
| dhrystone  | 354.60 DMIPS |867.00 DMIPS |+144%|

jserv committed:
Fix typo

jserv reviewed (changes requested):
Read [code refactoring](https://en.wikipedia.org/wiki/Code_refactoring) carefully.
> In computer programming and software design, code refactoring is the process of restructuring existing computer code — changing the factoring — without changing its external behavior.

However, the proposed pull request already changes the external behavior. Thus, don't use the term "refactor." You shall refine the subject.

jserv commented on src/riscv_private.h:
```diff
@@ -131,6 +131,7 @@ struct riscv_internal {
     block_map_t block_map; /**< basic block map */
 #else
     struct cache *cache;
+    struct cache *code_cache;
```
The variable name is misleading. Refine the two instances of `struct cache`.

jserv commented:
Known issue: JIT works on Linux/Arm64, but it fails on macOS/Arm64 (Apple M1).
```
$ make ENABLE_JIT=1 check
Running hello.elf ... [OK]
Running puzzle.elf ... Failed.
make: *** [check] Error 1
```

jserv commented:
Known issue: misalign test case fails.
```shell
$ make ENABLE_JIT=1 misalign 
Segmentation fault
Running aes.elf ... Failed.
```

qwe661234 committed:
Add CoreMark test script (#143)

The script calculates standard deviation and filters out outliers.

jserv committed:
Trivial clean up

qwe661234 committed:
Add missing EBB information in cbeqz and cbnez (#147)

Set last_pc in C.BEQZ and C.BEQZ instruction emulation.

Close #134

Co-authored-by: Yen-Fu Chen <qwe661234@gamil.com>

jserv commented:
I plan to merge this pull request as earlier as possible, so that we can rebase the recent `master` branch.
@qwe661234, it is acceptable to support x86-64 only at the moment. Refine the proposed changes as I mentioned.

qwe661234 committed:
Introduce baseline JIT compiler

The baseline JIT compiler design involves tracing an EBB when its usage
frequency exceeds a predetermined threshold. We then utilize a code generator
to convert the instruction sequence into C code. The C code is subsequently
compiled using MIR, and the resulting target machine code is stored in the
code cache for future utilization.

The primary objective of introducing the baseline JIT compiler is to enhance
the execution speed of RISC-V instructions with minimal modification. This
implementation requires three additional components: a code generator, MIR
compiler, and code cache, while maintaining the original design of the interpreter.
Furthermore, this baseline JIT compiler serves as the foundational target
for future improvements.

|  Metric   |        master           |       baseline JIT       |Speedup |
|-----------+-------------------------+--------------------------+--------|
| CoreMark  |1352.843 (Iterations/Sec)|1740.881 (Iterations/Sec) | +28.6% |
| dhrystone |1146 DMIPS               |2368.33 DMIPS             | +106%  |

Known issue: this implementation does not correctly work on Apple M1 MacOS.
da8ca3
